server/rooms.ts

```diff
@@ -1376,17 +1376,49 @@ export class GlobalRoomState {
 			// kill server in 10 seconds if it's still set to
 			setTimeout(() => {
 				if (Config.autolockdown && Rooms.global.lockdown === true) {
-					// finally kill the server
-					process.exit();
+					// finally save state / kill the server
+					void this.kill();
 				} else {
 					this.notifyRooms(
 						notifyPlaces,
-						`|html|<div class="broadcsat-red"><b>Automatic server lockdown kill canceled.</b><br /><br />In the last final seconds, the automatic lockdown was manually disabled.</div>`
+						`|html|<div class="broadcast-red"><b>Automatic server lockdown kill canceled.</b><br /><br />In the last final seconds, the automatic lockdown was manually disabled.</div>`
 					);
 				}
 			}, 10 * 1000);
 		}
 	}
+	async kill() {
+		if (Config.persistbattles) await this.writeBattleState();
+		return process.exit();
+	}
+	async writeBattleState() {
+		const buffer: AnyObject = {};
+		for (const [id, room] of Rooms.rooms) {
+			if (!room.battle) continue;
+			const formatid = room.battle.format;
+			if (!buffer[formatid]) buffer[formatid] = {};
+			const log = await room.battle.getLog();
+			if (!log) throw new Error(`Invalid battle log received while writing battle state.`);
+			buffer[formatid][id] = {
+				inputLog: log.join('\n'),
+				title: room.title,
+				roomid: room.roomid,
+			};
+		}
+		return FS(`logs/battles.json`).writeUpdate(() => JSON.stringify(buffer));
+	}
+	loadBattleState() {
+		if (!Config.persistbattles) return;
+		const battleData = JSON.parse(FS(`logs/battles.json`).readIfExistsSync() || "{}");
```

Comment:
Could I ask you to cast this somehow so it’s not `any` or whatever generic type `JSON.parse` returns? That would make it a lot easier for other people to debug this code later.